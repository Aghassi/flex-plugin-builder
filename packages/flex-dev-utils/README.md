![npm](https://img.shields.io/npm/v/flex-plugin-scripts.svg?style=flat-square)
![npm](https://img.shields.io/npm/dt/flex-plugin-scripts.svg?style=flat-square)
[![NpmLicense](https://img.shields.io/npm/l/flex-plugin-scripts.svg?style=flat-square)](LICENSE.md)

# Flex Plugin Scripts

Scripts to run, build, and deploy Flex plugin.

## Installation

This repo will automatically be installed as part of creating a new plugin.

## Scripts

### Release

```bash
npm run release {major, minor, patch, custom, overwrite} {--public}
```

This script will build, and then deploy your plugin bundle to Twilio Assets. Please note that your accountSid and authToken are required to use this script. Your plugins will be uploaded to Twilio Assets using the following format:

```text
https://domain.twil.io/plugins/${PluginName}/${Version}/bundle.js
```

Where `PluginName` is the `name` from `package.json`, and `Version` is the `version` from `package.json`. For example:

```text
https://domain.twil.io/plugins/dialpad/1.0.0/bundle.js
https://domain.twil.io/plugins/dialpad/1.1.0/bundle.js
https://domain.twil.io/plugins/dialpad/1.1.1/bundle.js`
```

You may choose between `major, minor, patch, custom, overwrite` options:

* **major**: A major SemVer release version, e.g. going from 1.2.3 to 2.0.0. 

    `npm run release major` 

* **minor**: A minor SemVer release version, e.g. going from 1.2.3 to 1.3.0. 

    `npm run release minor` 

* **patch**: A patch SemVer release version, e.g. going from 1.2.3 to 1.2.4.

    `npm run release patch`

* **custom**: A custom release version if you are not using SemVer versioning. Provide the version as the next argument.

    `npm run release custom v1.2-someTag`

* **overwrite**: Will overwrite the version. 

    `npm run release overwrite`

By default, all plugins are uploaded as Protected Assets (i.e. they require a Twilio Signature to be accessed). Plugin can be uploaded as a Public Asset using `--public` in your script.

### Build

```bash
npm run build
```

Run this script to build and create a single .js bundle for your plugin. You can then upload this to Twilio Assets or serve it via your own plugin service.

### Test

```bash
npm run build
```

Run this script to test your sourcecode.

### Start

```text
npm run start
```

Run this script to use dev-server for local development.

## Contributing

Make sure to follow the instructions in the [main repository](https://github.com/twilio/flex-plugin-builder#contributing) to set up the project.

```bash
# Install dependencies and link local packages with each other
cd packages/flex-plugin
npx lerna bootstrap

# Run tests
npm test

# To use your local package in a different project
npm link
# then in a different project
npm link flex-plugin
```

## Contributors

Thank you to all the lovely contributors to this project. Please check the main repository to see [all contributors](https://github.com/twilio/flex-plugin-builder#contributors).

## License

MIT
