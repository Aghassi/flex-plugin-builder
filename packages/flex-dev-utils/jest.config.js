const base = require('./../../jest.base.js');
const pkg = require('./package');

module.exports = {
  rootDir: '.',
  ...base(pkg),
  coverageThreshold: {
    global: {
      statements: 97,
      branches: 93,
      lines: 98,
      functions: 89,
    },
  },
};
