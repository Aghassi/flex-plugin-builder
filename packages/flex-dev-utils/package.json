{
  "name": "flex-dev-utils",
  "version": "3.0.2-alpha.0",
  "description": "Common development utility for creating a Flex plugin",
  "keywords": [
    "flex",
    "plugin",
    "twilio",
    "scripts"
  ],
  "homepage": "https://github.com/twilio/flex-plugin-builder#readme",
  "bugs": {
    "url": "https://github.com/twilio/flex-plugin-builder/issues"
  },
  "repository": {
    "type": "git",
    "url": "git+https://github.com/twilio/flex-plugin-builder.git"
  },
  "license": "MIT",
  "files": [
    "dist"
  ],
  "main": "dist/index.js",
  "types": "dist/index.d.ts",
  "scripts": {
    "prebuild": "rm -rf dist",
    "build": "tsc",
    "prepublish": "npm run build",
    "test": "cd ../.. && jest packages/flex-dev-utils",
    "test:watch": "cd ../.. && jest packages/flex-dev-utils --watch"
  },
  "dependencies": {
    "@types/boxen": "^3.0.1",
    "@types/log-symbols": "^3.0.0",
    "@types/update-notifier": "^2.5.0",
    "boxen": "^4.1.0",
    "chalk": "^2.4.2",
    "log-symbols": "^3.0.0",
    "ora": "^3.4.0",
    "update-notifier": "^3.0.0",
    "util": "^0.12.0"
  }
}
