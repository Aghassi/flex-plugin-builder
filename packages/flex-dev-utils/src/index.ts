export { default as fs } from './fs';
export { default as updateNotifier } from './updateNotifier';
export { default as ora } from './ora';
export { default as logger } from './logger';
