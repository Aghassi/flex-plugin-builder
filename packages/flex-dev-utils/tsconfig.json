{
  "extends": "../../tsconfig.json",
  "compilerOptions": {
    "outDir": "dist",
    "baseUrl": ".",
    "rootDir": "src",
    "esModuleInterop": true
  },
  "exclude": [
    "__mocks__",
    "node_modules",
    "dist"
  ]
}
