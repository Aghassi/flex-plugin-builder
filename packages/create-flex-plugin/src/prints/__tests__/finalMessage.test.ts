import { logger } from 'flex-dev-utils';
import * as boxed from 'flex-dev-utils/dist/boxen';

import finalMessage from '../finalMessage';
jest.mock('flex-dev-utils/dist/boxen');

describe('finalMessage', () => {
  beforeEach(() => {
    jest.resetAllMocks();
  });

  it('should render an npm setup message to the console', () => {
    let message = '';
    (boxed as any).info = jest.fn((msg) => message = msg);

    finalMessage({} as any);

    expect(boxed.info).toHaveBeenCalledTimes(1);
    expect(message).toMatchSnapshot();
  });

  it('should render a yarn setup message to the console', () => {
    let message = '';
    (boxed as any).info = jest.fn((msg) => message = msg);

    finalMessage({yarn: true} as any);

    expect(boxed.info).toHaveBeenCalledTimes(1);
    expect(message).toMatchSnapshot();
  });

  it('should render an instruction message skipping the setup step', () => {
    let message = '';
    (boxed as any).info = jest.fn((msg) => message = msg);

    finalMessage({install: true} as any);

    expect(boxed.info).toHaveBeenCalledTimes(1);
    expect(message).toMatchSnapshot();
  });
});
