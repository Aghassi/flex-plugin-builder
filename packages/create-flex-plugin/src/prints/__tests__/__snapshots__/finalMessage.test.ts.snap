// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`finalMessage should render a yarn setup message to the console 1`] = `
"Your Twilio Flex Plugin project has been successfully created!


[1m[32mSetup:[39m[22m
$ cd plugin-final-message/
$ yarn install

[1m[32mDevelopment:[39m[22m
$ cd plugin-final-message/
$ twilio flex:plugins:start

[1m[32mBuild Command:[39m[22m
$ cd plugin-final-message/
$ twilio flex:plugins:build

[1m[32mDeploy Command:[39m[22m
$ cd plugin-final-message/
$ twilio flex:plugins:deploy

For more info check the README.md file or go to:
âž¡ https://www.twilio.com/docs/flex"
`;

exports[`finalMessage should render an instruction message skipping the setup step 1`] = `
"Your Twilio Flex Plugin project has been successfully created!


[1m[32mDevelopment:[39m[22m
$ cd plugin-final-message/
$ twilio flex:plugins:start

[1m[32mBuild Command:[39m[22m
$ cd plugin-final-message/
$ twilio flex:plugins:build

[1m[32mDeploy Command:[39m[22m
$ cd plugin-final-message/
$ twilio flex:plugins:deploy

For more info check the README.md file or go to:
âž¡ https://www.twilio.com/docs/flex"
`;

exports[`finalMessage should render an npm setup message to the console 1`] = `
"Your Twilio Flex Plugin project has been successfully created!


[1m[32mSetup:[39m[22m
$ cd plugin-final-message/
$ npm install

[1m[32mDevelopment:[39m[22m
$ cd plugin-final-message/
$ twilio flex:plugins:start

[1m[32mBuild Command:[39m[22m
$ cd plugin-final-message/
$ twilio flex:plugins:build

[1m[32mDeploy Command:[39m[22m
$ cd plugin-final-message/
$ twilio flex:plugins:deploy

For more info check the README.md file or go to:
âž¡ https://www.twilio.com/docs/flex"
`;
